src/systems/isq/include/units/isq/thermodynamics.h

```diff
@@ -22,18 +22,77 @@
 
 #pragma once
 
-#include <units/dimension.h>
 #include <units/isq/base_quantities.h>
+#include <units/isq/mechanics.h>
 #include <units/isq/space_and_time.h>
+#include <units/quantity_spec.h>
 
 namespace units::isq {
 
-// inline constexpr struct thermodynamic_temperature : base_dimension<"Θ"> {} thermodynamic_temperature;
-// TODO Celsius temperature???
-
-// DERIVED_DIMENSION(mass_density, decltype(mass / volume));
-
-
-// DERIVED_DIMENSION(energy, decltype(force * length));  // defined in a mechanics header
+QUANTITY_SPEC(Celsius_temperature, thermodynamic_temperature);  // TODO should we account for T0 here?
```

Comment:
`T0` is the offset, which can be added to `quantity_spec`. That's the last point of https://github.com/mpusz/units/issues/405#issue-1468774884. That means, for differences, `1 °C = 1 K` and `1 F = ⁵⁄₉ K`, and when converting points, the offset is additionally taken into account.

Reply:
I am not so sure about it. I can create a `quantity` or `quantity_point` of `thermodynamic_temperature` and `Celsius_temperature` using either `K` or `deg_C` as unit. There is no way to restrict the usage of `K` to one and the usage of `deg_C` to other. Especially that ISO 80000 explicitly states that differences of `thermodynamic_temperature` can be expressed with either `K` or `deg_C`.

Reply:
> I am not so sure about it. I can create

This is strictly about conversions.

Reply:
I am still worried that there is no way to prevent the usage of `K` with `Celsius_temperature` or `deg_C` with `thermodynamic_temperature`.

Reply:
And there's no need to. As you comment, it's even standard endorsed.

I think I know what the source of your worry is. Many quantity equations in ISO-80000 do not adhere to the definition of [quantity equation](https://jcgm.bipm.org/vim/en/1.22.html). That includes the one for Celsius temperature, which is dependent on Kelvin.

Many of those quantities are correctly better represented in mp-units master and V2 as constants and unit constants, respectively. We can do a similar improvement of the representation. So while we could have a Celsius temperature quantity, perhaps it'd be better that the factor and offset are not part of the quantity, but the unit. Perhaps with a `quantity_spec` for units, `unit_spec`.

Reply:
Yes, this is exactly my concern. And yes, I feel that the offset for `deg_C` should be specified together with the unit and not with quantity, if only possible (this was exactly why I put this comment in the first place). I am not sure how to do it the best way, as `deg_C` is the only exception I am aware of here, so I wouldn't like to complicate or slow down all other cases.

Reply:
That's the price of modelling what we must. There's no escaping it.

Reply:
The "Remarks" says that only temperature differences may be expressed in either unit. The current design would allow expressing the "Kelvin temperature point 1 °C". Does that makes sense? What is its value as a Celsius temperature point?

Reply:
I am not sure what you mean by "Kelvin temperature point 1 °C". For `quantity`, both `thermodynamic_temperature` and `Celsius_temperature` can be expressed in either `K` or `deg_C`. The problem starts with `quantity_point`. If we somehow assign an offset to a `deg_C` unit then everything will work for both as well even though it would be strange. In this case, they will work as aliases.

Maybe we will decide to remove one of the `quantity_spec`s but let's wait for this decision.

Reply:
I don't think the "offset" should be part of the unit, but rather the quantity (this is the main beef I have with my favourite units library in python, [pint](https://pint.readthedocs.io/en/stable/), se my [comment here](https://github.com/hgrecco/pint/issues/1350)). The main argument is that, as soon as units appear in a compound expression, the offset immediately loses it's meaning (think for example the temperature coefficient of a resistor in kOhm/°C; obviously, that can only refer to temperature differences measured in °C, and there is simply no reasonable meaning to the "offset" of that quantity).

That said, it may make sense to define an implicit "offset" implied by the unit when describing a quantity point (e.g. a thermodynamic temperature) without explicitly specifying a reference point.

I also think we should really consider that information the "point of reference", not an "offset" - that one only appears between related quantities. But then, I have never really worked with quantity equations (and I don't have access to the ISO-80000 standard).

In the end, it would be great if the library would not only work for ISO thermodynamic temperatures, but also custom quantity points. Specifically, while the standard only endorses using °C to specify temperatures with respect to the "Celsisus temperature reference point", C++ programs may have a need to work with "unendorsed" combinations for efficiency reasons. As a specific example, we use this library on an embedded device employing a temperature sensor whose output is an 16 bit integer where zero corresponds to the temperature point -45 °C, and an increment of one corresponds to a temperature change of 0.01 °C. With #232, I can fully embed this information in the quantity_point and work with the original representation. However, I can also do computation with a floating point representation where the point of reference is still - 45 °C, but now I use SI °C for the scaling. Or, I can convert to the standard Celsius temperature reference point and continue to work with 0.01 °C increments, all in integer arithmetic.

Another use-case to consider is the glide-computer example, with different ways to express height/elevation, all using units of m and/or km.

Reply:
> I also think we should really consider that information the "point of reference", not an "offset"

That is more accurate.

> where zero corresponds to the temperature point -45 °C, and an increment of one corresponds to a temperature change of 0.01 °C.

So the formula of this quantity Q should be something like Q = 𝘵 × 0.01 - 45 °C, where 𝘵 is Celsius temperature. By specifying the factor of 0.01 to 𝘵, and the point of reference -45 °C, you should be able to do the same with this V2 framework.

Reply:
> That said, it may make sense to define an implicit "offset" implied by the unit when describing a quantity point (e.g. a thermodynamic temperature) without explicitly specifying a reference point.

In the current design, the `quantity_point` has access `point_origin`, `quantity_spec`, and a `unit`. In theory, any or all of them can introduce an offset for `quantity_point` arithmetic needs. It is up to us to specify what has sense and what does not. The problem with handling Celsius via `quantity_spec` is that with the current design, we cannot easily restrict a specific unit for a specific quantity. So both `thermodynamic_temperature` and `Celsius_temperature` can take `K` and `deg_C` as a unit. Even if we introduce some way to enforce such restrictions for those specific quantities similar problems will appear with all the derived quantities, like the mentioned temperature coefficient of a resistor in kOhm/°C.

Reply:
> In the end, it would be great if the library would not only work for ISO thermodynamic temperatures, but also custom quantity points.

Sure, I totally agree. This is what I used for testing but still didn't have enough time to make it an official example or unit test:

```cpp
  constexpr auto room_reference_temperature = quantity_point{21 * isq::Celsius_temperature[deg_C]};
  constexpr auto temperature_controller_delta = 3 * isq::Celsius_temperature[deg_C];

  constexpr quantity_point<isq::Celsius_temperature[deg_C], room_reference_temperature> room_default{0};
  constexpr quantity_point<isq::Celsius_temperature[deg_C], room_reference_temperature> room_low = room_default - temperature_controller_delta;
  constexpr quantity_point<isq::Celsius_temperature[deg_C], room_reference_temperature> room_high = room_default + temperature_controller_delta;

  static_assert(room_low.relative() == -3 * isq::Celsius_temperature[deg_C]);
  static_assert(room_low.relative() == -3 * isq::thermodynamic_temperature[K]);
  static_assert(room_default.relative() == 0 * isq::Celsius_temperature[deg_C]);
  static_assert(room_default.relative() == 0 * isq::thermodynamic_temperature[K]);
  static_assert(room_high.relative() == 3 * isq::Celsius_temperature[deg_C]);
  static_assert(room_high.relative() == 3 * isq::thermodynamic_temperature[K]);

  static_assert(room_low.absolute() == 18 * isq::Celsius_temperature[deg_C]);
  // static_assert(room_low.absolute() == (273.15 + 18) * isq::thermodynamic_temperature[K]);
  static_assert(room_default.absolute() == 21 * isq::Celsius_temperature[deg_C]);
  // static_assert(room_default.absolute() == (273.15 + 21) * isq::thermodynamic_temperature[K]);
  static_assert(room_high.absolute() == 24 * isq::Celsius_temperature[deg_C]);
  // static_assert(room_high.absolute() == (273.15 + 24) * isq::thermodynamic_temperature[K]);

  static_assert(room_high - room_low == 6 * isq::Celsius_temperature[deg_C]);
  static_assert(room_high - room_low == 6 * isq::thermodynamic_temperature[K]);
  static_assert((room_high - room_low)[K] == 6 * isq::Celsius_temperature[deg_C]);
```

The commented out code still does not work but I think it should.

Reply:
Another example I prepared while working on V2 `quantity_point` was:

```cpp
  struct mean_sea_level : absolute_point_origin<isq::height> {};

  constexpr quantity_point<isq::height[m], mean_sea_level{}> ZRH_ground_level{432 * isq::height[m]};
  constexpr quantity_point<isq::height[m], mean_sea_level{}> GDN_ground_level{149 * isq::height[m]};
  constexpr quantity_point<isq::height[m], ZRH_ground_level> ZRH_afe{200 * isq::height[m]};
  constexpr quantity_point<isq::height[m], GDN_ground_level> GDN_afe{450 * isq::height[m]};
  static_assert(ZRH_afe.absolute() == 632 * isq::height[m]);
  static_assert(GDN_afe.absolute() == 599 * isq::height[m]);
  static_assert(GDN_afe - ZRH_afe == -33 * isq::height[m]);
  static_assert(GDN_ground_level - ZRH_ground_level == -283 * isq::height[m]);
  static_assert(ZRH_afe - GDN_ground_level == 483 * isq::height[m]);
```

Reply:
@burnpanck, please let me know if that is what you meant, and if not, how we can improve here?

Reply:
That looks all pretty cool to me, and I especially like directly specifying the reference point as quantity_point NTTP template argument, e.g. as in `constexpr quantity_point<isq::height[m], ZRH_ground_level>`.

The only place where I tend to disagree are the `.absolute()` accessors. To me, I lack a clear explanation what that would semantically do. `q.relative()` is easy: It is the same as `q - q.origin_point`. `q.absolute()` doesn't mean anything to me. A potential definition would be `q.absolute() == q - q.ultimate_origin_point`, and we could reasonably require that `ultimate_origin_point` is one arbitrary point that is the same for all quantity_points that are interconvertible. While this doesn't fully specify `ultimate_origin_point`, the only reasonable choice for thermodynamic temperatures would be (thermodynamic) absolute zero, so we would have your commented out variants correct instead, and the currently passing static_asserts would fail instead: `room_low.absolute() == isq::thermodynamic_temperature(237.15+18, K) == isq::Celsius_temperature(237.15+18, deg_C)`. That said, we could also just leave `.absolute` out of the initial interface. (In fact, given the `q.relative() == q - q.origin_point` equivalence, we could also leave the `.relative` out of the interface).

Why cannot we have both `X == isq::Celsius_temperature(18, deg_C)` and `X == isq::thermodynamic_temperature(273.15 + 18, K)`, for any hypothetical `X`? It is the same argument why we cannot have `Y == isq::radius(2, m)` and `Y == isq::diameter(1, m)` for any hypothetical `Y`: In both cases, we may interpret the ISO 80000 standard for the two quantities to be "equal". However, as we have seen, it turns out that this is at odds with the mathematical rules of standard operations when applied between quantities that aren't of equal <kind> (again; we lack a good name to to describe that equivalence class/concept): If we *define* `isq::radius(2, m) == isq::diameter(1, m)`, then there will be no reasonable definition of `isq::radius(2, m) + isq::length(1, m)` that will preserve associativity and commutativity - that being unacceptable, the only option is to disallow it completely. The exact same happens for the temperatures: If we *define* `isq::Celsius_temperature(0, deg_C) == isq::thermodynamic_temperature(273.15, K)`, then there is no reasonable definition of `2*(isq::Celsius_temperature(0, deg_C) + isq::thermodynamic_temperature(0, K))` (unless you consider `2*isq::Celsius_temperature(0, deg_C) == isq::Celsius_temperature(273.15, deg_C)` reasonable). Furthermore, there is no reasonable definition to `isq::thermodynamic_temperature(5, K) + imperial::Fahrenheit_temperature(9, deg_F)`.

For ISO 80000, mixed quantity operations aren't an issue, as it doesn't talk about mathematical operations between quantities at all as far as I know (though I don't know ISO 80000 well). For a library intent on modelling physical relations, it is however. This is why essentially all quantity libraries in all languages I know of discard those special equalities in their "normal mode" and sometimes provide special ways to opt-in into those equalities (e.g. astropy's equivalencies, mp-units' quantity_points, etc...).